### ISSUE TITLE
HTTP/1.1, HTTPS, HTTP/2 Forward proxy concepts

### ISSUE THREAD

**Issue by aafanasyev** [nghttp2/nghttp2#547]
Hi,

I think I have not clear understanding of the Forward Proxy in nghttpx. As I read it here:
http://stackoverflow.com/questions/516323/https-connections-over-proxy-servers
The forward proxy with features like caching, content and url filtering is almost not possible especially with built-in certificate pins. Looks like it kills idea of forward proxy with features above completely. 

According https://github.com/tatsuhiro-t/nghttp2#nghttpx---proxy 
nghttpx implements important performance-oriented features in TLS, such as session IDs, session tickets (with automatic key rotation), OCSP stapling, dynamic record sizing, ALPN/NPN, forward secrecy and SPDY & HTTP/2.  nghttpx  also offers the functionality to share session cache and ticket keys among multiple  nghttpx  instances via memcached.
As I see nghttpx plays actually as some kind of gateway (tunnel) for clients that would like directly, connect to a HTTP/1.1, HTTPS, HTTP/2 webserver, or security proxy (HTTP traffic from server/proxy to nghttpx http/2 traffic to server. For me is more interesting tunnel mode for now. Because as I see it looks more like classical forward proxy(IMHO). Original protocol comes in and comes out. How can I do URL or content filtering in such mode? Memcache and nghttpx. Is it about caching traffic? Is any example around?

Thank you.

**Comment by tatsuhiro-t** [comment]
Firstly, nghttpx does not cache a thing at all.  When configured as forward proxy (-s option), it requires HTTP proxy behind the nghttpx.  The backend HTTP proxy may have caching ability.  We have no plan to implement MITM proxy feature in nghttpx.  So it is not possible to filter end-to-end encrypted traffic in nghttpx.

**Comment by aafanasyev** [comment]
Oh great! Thank you for your fast response. 

As I understood, I can configure nghttpx without http proxy and if nghttpx if have no http proxy behind nghttp can play only as a proxy that hides clients?  Am I right? Thank you.

**Comment by tatsuhiro-t** [comment]
If you'd like to use nghttpx as forward proxy, you need backend HTTP proxy.

**Comment by aafanasyev** [comment]
Hm, 
So it is not possible to configure it like this:
Client <-- (HTTP/2) --> nghttpx <-- (HTTP/2) --> nginx
                                  [secure proxy]                  [web server]
or like this

Client <-- (HTTP/1.1) --> nghttpx <-- (HTTP/1.1) --> nginx
                                  [secure proxy]                 [web server]

Thank you.

**Comment by tatsuhiro-t** [comment]
Yes.  I think this is typical reverse proxy setup.

**Comment by aafanasyev** [comment]
Sorry, I am confused... In scheme above the nghttpx facing the internet. I thought the nghttpx in forward mode capable of one or other functionality. 

- Caching responses from web server(not implemented in nghttpx), 
- Forward proxy can block client request (ACL), 
- Hide clients, 
- Tunnel client requests (HTTPS), 
- Decrypt the TLS and send plain data to the client (SSL bump/SSL Peek and Splice). 

Please correct me if I am wrong!
According documentation, I do see the nghttpx in foward mode have only functions like: 
-Forward requests, 
-Hide clients 
-Secure communication in insecure closed subnets (http/1.1 -> http/2||https)
-Decrypt secure traffic and forward clear traffic to the client in a secure closed subnet(http/2||https -> http/1.1).

Is that correct? Do I extrapolate this information correct?

P.S.: Please sorry me for stupid questions. I worked mostly with Squid and functionality like caching/forwading/blocking traffic. Nowadays, only nghttpx have developed forward proxy mode (according docs), but it is not so clear for me how it is working in this mode.

Thank you for you patient.

**Comment by tatsuhiro-t** [comment]
Ah, I was misunderstood your comment.
I think your understanding is correct.  Just think that nghttpx is not complete forward proxy as its own.  It rather the frontend of another forward proxy, and nghttpx does the TLS termination and protocol translation (e.g., HTTP/2 -> HTTP/1.1).  And it lacks caching.
Therefore we need another forward proxy as the backend server behind nghttpx.
In other words, when squid implements HTTP/2, then you don't have to nghttpx anymore, can just use squid instead.

**Comment by pnip** [comment]
@tatsuhiro-t Does nghttpx handles SSL-bumped, specifically if a client send a CONNECT to it does it know to upgrade the underlying connection to SSL?

**Comment by tatsuhiro-t** [comment]
nghttpx does not care what is transmitted after successful CONNECT.

**Comment by yaung** [comment]
Hello ! 
> Hm,
>So it is not possible to configure it like this:
>Client <-- (HTTP/2) --> nghttpx <-- (HTTP/2) --> nginx
>[secure proxy] [web server]

I want to kown did you achieve this purpose ? Thank you !

**Comment by aafanasyev** [comment]
Not really. I do not have sources now, but I remember it was falling multiple time.
